### ISSUE TITLE
Runtime event for Monitor.Wait

### ISSUE THREAD

**Issue by verdie-g** [dotnet/runtime#94264]
<div type='discussions-op-text'>

<sup>Originally posted by **verdie-g** August 17, 2023 https://github.com/dotnet/runtime/discussions/90740</sup>
I recently got a latency spikes issue in my application because the async API of a library was actually waiting on a ManualResetEventSlim.Wait that would cause thread starvation. Now I would like to collect events about any thread that starts waiting. I can collect the contention event, the task waits, but it seems like there is no events emitted in ManualResetEventSlim. Looking at the code [ManualResetEventSlim uses Monitor.Wait](https://github.com/dotnet/runtime/blob/45acd380b37c9ee883070a70a2ef2cb7eca77683/src/libraries/System.Private.CoreLib/src/System/Threading/ManualResetEventSlim.cs#L561) so I've tried looking for an event there but couldn't find any. Any idea what I could do to easily detect blocked threads?</div>

What do you think of having a new event for Monitor.Wait?

Related issue where I suggested a dedicated event for synchronous task waits: https://github.com/dotnet/runtime/issues/63016.

**Comment by verdie-g** [comment]
@noahfalk what do you think about that?

In my experience, most latency issues come from thread pool threads being blocked. It can be by a lock, by sync-over-async or simply by plain synchronous operations (e.g. [HttpClient.Send](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.send)).

When I see the `monitor-lock-contention-count` event counter increasing, I know that I can start collecting `ContentionStart` to get stack traces about where that happened. 

When `monitor-lock-contention-count` is low but I still see spikes of `threadpool-thread-count` or `threadpool-queue-length` I can try dumping the thread stacks and hoping to find the culprit but it's very likely I won't find anything.

If `Monitor.Wait` was emitting an event I could quickly find who's responsible for blocking the thread.

**Comment by noahfalk** [comment]
When you say add an event I assume you are talking about an EventSource event? 

Off the cuff this seems like a reasonable thing to have, though I haven't investigated any details about hooking into the monitor implementation. First few thoughts:
 - There is recursion/deadlock risk if someone hooked up an EventListener, but it seems no worse than what already exists on the ContentionStart event. I feel OK telling people that if they want to hook callbacks into low-level synchronization mechanisms then they need to be careful what they do inside of that callback.
 - Similar to contention start we need to be cautious about the perf impact, but if we are only triggering the event where we are about to do a context switch or Task switch anyways I'd certainly hope the IsEnabled() check is unnoticeable. I'd expect the event to be off by default.
 - There are other sample-based ways to diagnose such as using dotnet-stack or dotnet-trace "cpu-sampling" (its not accurately named). Those approaches can show the callstacks which are keeping threads busy, but they only show the stacks at the moment the sample occurs - the standard tradeoff between sampled and instrumented profiling. I don't see any reason that having one should preclude us from offering the other.
 
 @kouvel - Any thoughts about doing this? 
 @verdie-g - Were you interested in investigating further and making a PR? Its fine if you just wanted to make the request only, just more at mercy of competing work that way.

**Comment by verdie-g** [comment]
> When you say add an event I assume you are talking about an EventSource event?

I was thinking about an event like `ContentionStart`. I thought that EventSource events were things like `System.Diagnostics.Tracing.PollingCounter` but I understand now that non-counter events are also from emitted from EventSource.

Btw, it could also be interested to have a counter for the Monitor.Wait, a spike of this counter could point out an issue.

> Were you interested in investigating further and making a PR

I would love to. I'll start working on it, it will take some time though. I think I can mimick how the `ContentionStart` is emitted.

**Comment by danmoseley** [comment]
How do we think about EventSource events going forward (vs open telemetry)

**Comment by verdie-g** [comment]
Good point, I understood you were moving away from EventSource counters in favor of OpenTelemetry instruments. Though for punctual events such as `ContentionStart` I don't think there is an equivalent in OpenTelemetry. It's too granular for a span/activity. So I would suppose EventSource events are still the way to go for such use-case?

**Comment by verdie-g** [comment]
Drafted https://github.com/dotnet/runtime/pull/94737

**Comment by danmoseley** [comment]
Cc @tarekgh @JamesNK also, for the eventsource/otel question..

**Comment by tarekgh** [comment]
Can OTel [contentioncount](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.Runtime/README.md#processruntimedotnetmonitorlock_contentioncount) help in this scenario?

**Comment by verdie-g** [comment]
> Can OTel [contentioncount](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Instrumentation.Runtime/README.md#processruntimedotnetmonitorlock_contentioncount) help in this scenario?

This count is based on [Monitor.LockContentionCount](https://learn.microsoft.com/en-us/dotnet/api/system.threading.monitor.lockcontentioncount) which is a different use-case (see my previous message https://github.com/dotnet/runtime/issues/94264#issuecomment-1806898515).

**Comment by noahfalk** [comment]
> How do we think about EventSource events going forward (vs open telemetry)

I would distinguish two parts of the scenario:
  (1) implementing the telemetry using the canonical telemetry APIs that we often use and talk about with OpenTelemetry (ILogger, Meter, Activity)
  (2) being able to egress this data via OpenTelemetry SDK somehow.
  
For anything that is happening in the native and low-level managed portions of the stack none of those standard APIs we use in (1) are available. EventSource or C# delegate callbacks are pretty much the only option at the layer Monitor.Wait operates at. Even if we did have the ability to change our layering and make something like ILogger available, a synchronous callback inside Monitor.Wait() probably can't run the typical code ILoggerProviders use without causing deadlocks. Its probably for the best that this event would be EventSource only.

For (2) I can certainly imagine a future scenario where OpenTelemetry SDK would hook into EventPipe allowing low level events to queued up, serialized, and merged with other outgoing log telemetry. We've already got other events that are even harder to handle synchronously than the proposed Monitor.Wait event such as GC events that are triggered when it is illegal for managed code to run. I don't think there we would need to tie delivering that future scenario to making initial progress on the EventSource part though.

> Good point, I understood you were moving away from EventCounters in favor of OpenTelemetry instruments

Thats true, but separate from the EventSource event. For example if you wanted both an EventSource event and a S.D.Metrics.Counter instrument I would suggest writing the code as:
- define a new public static property in CoreLib that is incremented each time monitor.wait occurs
- if the EventSource event is enabled, trigger the event
- In some higher level library (System.DiagnosticSource perhaps?), define a new ObservableCounter instrument implemented with a callback that reads that new static property we defined.

> Drafted https://github.com/dotnet/runtime/pull/94737

That was fast! I haven't had a chance to look at it just yet, but its on my todo list. I do think we should get feedback from @kouvel before you spend too much time here.

**Comment by verdie-g** [comment]
> define a new public static property in CoreLib that is incremented each time monitor.wait occurs

Actually I would like to add such counter in addition to the event in https://github.com/dotnet/runtime/pull/94737. 

Regarding OTEL VS EventCounters, from https://github.com/dotnet/runtime/issues/85372 it seems like it's still unclear if current EventCounters will be turned into OTEL instruments. So I think we should:
1. Create that `public static long WaitCount` (as you suggested)
2. Create a `monitor-wait-count` EventCounter alongside [`monitor-lock-contention-count`](https://github.com/dotnet/runtime/blob/2a9c80b799c55cba30c755d60042a6acb30aa37f/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/RuntimeEventSource.cs#L108) for consistency
3. Extend [opentelemetry-dotnet-contrib](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/ab6398b4155bcd09424520f2e7aebdb0354f43b1/src/OpenTelemetry.Instrumentation.Runtime/RuntimeMetrics.cs) to include that new metric

Should I open an API proposal to discuss that point or do you think I could go straight to the PR?

**Comment by verdie-g** [comment]
> Should I open an API proposal to discuss that point or do you think I could go straight to the PR?

I had some time so I drafted https://github.com/dotnet/runtime/pull/94810

**Comment by kouvel** [comment]
> I recently got a latency spikes issue in my application because the async API of a library was actually waiting on a ManualResetEventSlim.Wait that would cause thread starvation.

> In my experience, most latency issues come from thread pool threads being blocked. It can be by a lock, by sync-over-async or simply by plain synchronous operations (e.g. [HttpClient.Send](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.send)).

> Any idea what I could do to easily detect blocked threads?

In cases where several thread pool threads are getting blocked and leading to issues (maybe even all thread pool threads are getting blocked, leading to thread pool starvation), the time spent blocking is typically long-enough that they can be relatively easily caught with a dotnet-trace sampling profile. The ThreadTime view in PerfView seems to do a somewhat decent job of isolating blocking stacks.

Events can be useful in other cases, some examples:
- When the wait durations are short-enough that a sampling-based approach may not accurately reflect the cumulative time spent waiting on a particular primitive. For example, when contention counts are spiking in a scenario, it may not be a starvation-type situation, and collecting the ContentionStart/End events in the scenario would help to identify which locks are being waited upon most frequently or for the longest duration, along with the wait stack traces.
- When more specific information can be included to help with diagnostics. For example the ContentionStart/End events can now also be used to identify which thread is holding the lock, enabling getting that thread's stack trace at that moment in time.

I'm interested in understanding in what kind of cases adding events would provide better support for diagnostics than a sampling-based approach.

> @kouvel - Any thoughts about doing this?

Probably most waits on synchronization primitives could be distilled into waiting on a WaitHandle. Instead of adding an event for each synchronization primitive / use case, a better approach may be to add events for the underlying WaitHandle wait (the implementation of which is in different places depending on the runtime being used). When the event is enabled, it may also be necessary to do a zero-timeout wait before emitting events, to weed out false-positives and also reduce noisiness of the events.

There are other types of waits that are common and not on synchronization primitives, I imagine it would be difficult and perhaps unnecessary to catch all of them, and some of those events could be noisy enough to affect the perf characteristics of the app, which can make it difficult to collect the info in production environments. I think it would be reasonable to add events when they could provide additional information, then it can be determined how to do so.

**Comment by verdie-g** [comment]
> I'm interested in understanding in what kind of cases adding events would provide better support for diagnostics than a sampling-based approach.

The problem that originally made me open that issue was a service that was refreshing synchronously (blocking the thread) a value from a database whenever it would receive an invalidation message. Most of the time the number of invalidation was low so the impact could not be observed. However, once a day I could observe a huge spike in the number of thread and it took me a lot of time to find the root cause.

With a 1 minute sampling profile, the number of invalidations would be too low to appear in the results. With a X hours sampling profile, it would still be hard to see the spike of invalidations among all the other data.

If I had an event counter for `Monitor.Wait` I could have directly known what to look for. And if `Monitor.Wait` was emitting an event, with a 1 minute event collection, just a single event emitted from the thread pool would have pointed out something wrong in the code.

**Comment by kouvel** [comment]
I see, thanks for the info. Indeed when it's difficult to catch the issue in action, other info like events/counters may offer clues.

> However, once a day I could observe a huge spike in the number of thread

> With a X hours sampling profile, it would still be hard to see the spike of invalidations among all the other data.

If a spike is caught in the profile, assuming that it's causing some starvation-like issues, I'd expect that it would be visible, perhaps more easily in the ThreadTime view in PerfView. In that type of case a long-duration profile taken with even a reduced sampling frequency may be sufficient to catch the issue.

> If I had an event counter for Monitor.Wait I could have directly known what to look for. And if Monitor.Wait was emitting an event, with a 1 minute event collection, just a single event emitted from the thread pool would have pointed out something wrong in the code.

Were the problematic calls to Monitor.Wait coming from waits on Tasks? I can see the benefit of adding a counter here, but if it's from waiting on Tasks, it would probably be more meaningful to add a counter for Task waits, perhaps in FrameworkEventSource, rather than for Monitor.Wait, which is a lower-level primitive that can be used for a variety of things. Similarly, the thread pool compensates for waits on Tasks by injecting threads more quickly, but not for Monitor.Wait because there can easily be cases in the latter bucket that would be counterproductive to compensate for. If it is due to waits on Tasks, I'd suggest starting with filing an issue in System.Threading.Tasks to see if a counter can be added.

It sounds like there was only one case in your scenario where Monitor.Wait is called most frequently, pointing directly to the problem even though it's not yet causing an issue in the perf metrics. While that's of course possible, the usage of that info to identify a problem in the general variety of cases can be very hit-or-miss.

As for events, it seems like an event raised for waiting on a WaitHandle as I mentioned before would similarly suffice in your scenario (the stacks can easily be filtered to those caused by Monitor.Wait). It seems like that could be useful to have in certain situations. If that would work for you, my suggestion would be to file a separate issue that targets that specific request, and perhaps this issue could track the whole.

**Comment by verdie-g** [comment]
> If it is due to waits on Tasks, I'd suggest starting with filing an issue in System.Threading.Tasks to see if a counter can be added.

For that specific issue it was indeed a `Task.Wait()` ([more info here](https://datastax-oss.atlassian.net/browse/CSHARP-994)). I've actually already open an issue about a task wait counter two years ago https://github.com/dotnet/runtime/issues/63016 but it was not convincing. I thought that having an event at a lower level would be easier to justify.

I had a previous issue where the load was a `HttpClient.Send` but I just checked and the implementation seems to be sync-over-async so it should be caught by an event in Task.Wait.
![image](https://github.com/dotnet/runtime/assets/9092290/9199e8aa-0714-46f1-bca2-842faefbab48)

> it seems like an event raised for waiting on a WaitHandle as I mentioned before would similarly suffice

That wouldn't catch Task.Wait right? I'm not sure what public api uses WaitHandle actually.

Btw there is already an event in `Task.Wait`, I mention it in https://github.com/dotnet/runtime/issues/63016. One problem with it is that it's super spammy because it contains async waits too. It's still an event I use regularly, I'm not sure if I just misread the event collection in my original problem or if I just did not collect that event.

**Comment by kouvel** [comment]
> That wouldn't catch Task.Wait right? I'm not sure what public api uses WaitHandle actually.

It believe it would, Task.Wait uses ManualResetEventSlim.Wait, which in turn uses Monitor.Wait, which in turn waits on a wait handle (in coreclr's implementation, though it's likely it would be similar in other runtimes).

> One problem with it is that it's super spammy because it contains async waits too.

Yea it seems it is raised even before spin-waiting, that would be way too noisy. Perhaps it could be adjusted to be more useful, I don't know if there is a good reason for why it is raised as early as it is.

**Comment by verdie-g** [comment]
> Task.Wait uses ManualResetEventSlim.Wait, which in turn uses Monitor.Wait, which in turn waits on a wait handle

Could you please share some pointer on that wait handle. WaitHandle seems to be a managed class but Monitor.Wait transitions to native code.

**Comment by kouvel** [comment]
> Could you please share some pointer on that wait handle.

The managed version of WaitHandle and derived types are very simple wrappers around OS primitives. On Windows, WaitHandle types like Mutex, Semaphore, etc., map directly to the OS primitives, while on Unixes and on other runtimes they map to some sort of PAL that serves as a wait subsystem that behaves somewhat equivalently.

> WaitHandle seems to be a managed class but Monitor.Wait transitions to native code.

Wait handles are a Windows concept that are emulated by coreclr on Unixes by the PAL. The managed WaitHandle implementations are thin wrappers around OS primitives (Windows) or PAL counterparts (Unixes).

The underlying implementation of WaitHandle in any runtime is likely to transition into native code, perhaps with at least a p/invoke into some native primitive.

For Monitor.Wait, following down the path from the call to [SyncBlock::Wait](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/syncblk.cpp#L2873), it seems to eventually end up waiting on a wait handle.

**Comment by verdie-g** [comment]
So for the CoreCLR Monitor.Wait, the call chain looks like:
[ObjectNative::WaitTimeout](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs#L156)
[ObjHeader::Wait](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/object.h#L337)
[SyncBlock::Wait](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/syncblk.cpp#L2261)
[Thread::Block](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/syncblk.cpp#L2873)
[CLREventBase::Wait](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/threads.cpp#L3945)

if alertable
[Thread:DoAppropriateWait](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/synch.cpp#L459)
[Call_RetI4](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/threads.cpp#L3884)
else
[WaitForSingleObjectEx](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/synch.cpp#L372)

I suppose there is a PAL behind both these leaf functions. Where would you expect the event to be emitted?

Note that for NativeAot the implementation seems to be C# only.

**Comment by kouvel** [comment]
`CLREventBase::Wait` also eventually ends up in `Thread::DoAppropriateWait`, so I was thinking that might be a good place for an event.

Thinking about it a bit more, a disadvantage of raising the event there is that the relevant managed object is not immediately available. It would probably be useful to have it available in the event in case someone would want to take a heap dump of the process along with collecting events to look at the relevant objects involved. Maybe the object can be plumbed through in the `PendingSync *syncState` parameter.

For NativeAOT/Mono, it would probably be sufficient to include the WaitHandle object in the event, looking at GC references to it in a heap dump could tell what are the relevant objects.

**Comment by kouvel** [comment]
Maybe it would also be useful to include an enum parameter in the event indicating what kind of wait it is. Calls to `WaitHandle.WaitOne()` in managed implementations could use an `Unknown` value, and `Monitor.Wait` could use an internal overload that includes a specific value.

**Comment by verdie-g** [comment]
> CLREventBase::Wait also eventually ends up in Thread::DoAppropriateWait(int, ...), so I was thinking that might be a good place for an event.

I'll try emitting the event from CLREventBase though I'm not sure if I'll be able to get the object reference from there. Let's see.

> Maybe it would also be useful to include an enum parameter in the event indicating what kind of wait it is

I'm not sure what info you want here. Is it managed/native like [ContentionFlagsMap](https://github.com/dotnet/runtime/blob/7f42ffa3785dcb707f2ee9365f09e0058f21482d/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/NativeRuntimeEventSource.Threading.NativeSinks.cs#L60)?

**Comment by kouvel** [comment]
> I'll try emitting the event from CLREventBase

`Thread::DoAppropriateWait(int, ...)` may be a better place for the event. That's where the Windows/PAL Wait function is called, and it would be possible to do a zero-timeout wait first when the event is enabled. Waits overridden by SyncrhonizationContext also typically end up there.

**Comment by kouvel** [comment]
> I'm not sure what info you want here. Is it managed/native like [ContentionFlagsMap](https://github.com/dotnet/runtime/blob/7f42ffa3785dcb707f2ee9365f09e0058f21482d/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/NativeRuntimeEventSource.Threading.NativeSinks.cs#L60)?

I was thinking something like this:
https://github.com/dotnet/runtime/blob/7f42ffa3785dcb707f2ee9365f09e0058f21482d/src/coreclr/vm/ClrEtwAll.man#L512-L521

Except the enum values could be `Unknown, MonitorWait`, and things could be added onto that in the future.

**Comment by verdie-g** [comment]
> Thread::DoAppropriateWait(int, ...)

But see [here](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/synch.cpp#L457-L467) only of of the two branches will call `DoAppropriateWait`.

> Except the enum values could be Unknown, MonitorWait, and things could be added onto that in the future.

Shouldn't the developer rely on the StackTrace for that? Or maybe you have a specific use-case in mind where filtering/aggregating by this enum would be interesting? Btw, is MonitorWait the method that can end up in this code?

**Comment by kouvel** [comment]
> But see [here](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/synch.cpp#L457-L467) only of of the two branches will call DoAppropriateWait.

`Monitor.Wait` and most managed waits are alertable waits, so it's probably sufficient for now to add the event for alertable cases.

>  where filtering/aggregating by this enum would be interesting?

Yes that's what I was thinking. It also makes it easy to scan over a large number of the events without having to look at the stack trace for each interesting one.

> Btw, is MonitorWait the method that can end up in this code?

Yes, here's the call chain:
- [Monitor.Wait](https://github.com/dotnet/runtime/blob/7f42ffa3785dcb707f2ee9365f09e0058f21482d/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs#L151)
- [Monitor.ObjWait](https://github.com/dotnet/runtime/blob/7f42ffa3785dcb707f2ee9365f09e0058f21482d/src/coreclr/System.Private.CoreLib/src/System/Threading/Monitor.CoreCLR.cs#L148)
- Monitor.ObjWait [maps to](https://github.com/dotnet/runtime/blob/7f42ffa3785dcb707f2ee9365f09e0058f21482d/src/coreclr/vm/ecalllist.h#L539) `ObjectNative::WaitTimeout`

And the rest is what you posted above.

**Comment by verdie-g** [comment]
Missed a word, I meant "is MonitorWait the **only** method that can end up in this code?". I assume so according to your last message.

> Monitor.Wait and most waits from managed APIs are alertable waits, so it's probably sufficient for now to add the event for alertable cases.

Got it. I'll update the PR.

**Comment by kouvel** [comment]
> Missed a word, I meant "is MonitorWait the only method that can end up in this code?". I assume so according to your last message.

Ah ok I misunderstood. Most waits on synchronization primitives from managed APIs would end up here.

**Comment by verdie-g** [comment]
I don't think we'll be able to get any relevant info for the event in [Thread::DoAppropriateWait](https://github.com/dotnet/runtime/blob/33d7f9815a4cf0ceb19f7014f978137218855a32/src/coreclr/vm/threads.cpp#L3154). The method just receives an array of handles so we can't get the object pointer or the lock id. Are you fine with that or should we think of an alternative?

**Comment by kouvel** [comment]
Can the info be passed in through the `syncState` parameter?

**Comment by kouvel** [comment]
The object is available in `ObjectNative::WaitTimeout`, it could be passed through.

**Comment by verdie-g** [comment]
> The object is available in ObjectNative::WaitTimeout, it could be passed through.

Sure but why not emit the event in `ObjectNative::WaitTimeout` in that case? If `Thread::DoAppropriateWait` takes an array I'm pretty sure we'll find callers of that method that can't pass a single object.

**Comment by kouvel** [comment]
> Sure but why not emit the event in `ObjectNative::WaitTimeout` in that case? If `Thread::DoAppropriateWait` takes an array I'm pretty sure we'll find callers of that method that can't pass a single object.

Single and multi waits end up in `Thread::DoAppropriateWait`, the event could be raised only when it's a wait one a single wait handle. I figured this type of info come in useful in other cases too and that would be a common place to raise one event that captures that type of info, instead of having a separate event for each case.

**Comment by verdie-g** [comment]
Got it. What do you think of always emitting the event, but passing null for the object pointer for multi-waits?

Other question, @noahfalk was suggesting `MonitorWait` for the event name. Are you fine with that?

**Comment by kouvel** [comment]
> What do you think of always emitting the event, but passing null for the object pointer for multi-waits?

That would probably be fine too.

> Other question, @noahfalk was suggesting `MonitorWait` for the event name. Are you fine with that?

I guess if the event were specific to Monitor.Wait, then MonitorWaitStart/End would make sense. Since we're talking about emitting a general event that could be used for other things, maybe WaitHandleWaitStart/End would work?

**Comment by verdie-g** [comment]
I've updated #94737. @kouvel could you review have a quick look to validate the direction taken and I'll complete the PR with the managed WaitHandle, aot, mono...

**Comment by kouvel** [comment]
I've added a few comments there, looks like a good start to me.

**Comment by verdie-g** [comment]
I'll:
1. add the wait source to perfview any stack https://github.com/microsoft/perfview/pull/1958
2. update our runtime
3. collect the event on a prod service
before closing the issue.

**Comment by verdie-g** [comment]
🎉 
![image](https://github.com/dotnet/runtime/assets/9092290/d50aa875-b17f-4ca7-ac3b-c2359e5261e7)
